emcc.py

```diff
@@ -3450,26 +3445,32 @@ def modularize():
         'src': src
       }
   else:
-    # Create the MODULARIZE_INSTANCE instance
-    # Note that we notice the global Module object, just like in normal
-    # non-MODULARIZE mode (while MODULARIZE has the user create the instances,
-    # and the user can decide whether to use Module there or something
-    # else etc.).
+    promise_variable_suffix = '_promise'
+    # Create the promise for the MODULARIZE_INSTANCE instance. The name of the
+    # variable will be `Module_promise` by default. If the EXPORT_NAME option
+    # is specified, the variable named will be `{EXPORT_NAME}_promise`.
+    # 
+    # Note that the global, custom-named `Module` object is passed to the
+    # factory function invocation, just like in non-MODULARIZE mode. This is
+    # different than code built with the MODULARIZE option where the user
+    # specifies any Module seed when calling the factory function themselves.
     src = '''
-var %(EXPORT_NAME)s = (%(src)s)(typeof %(EXPORT_NAME)s === 'object' ? %(EXPORT_NAME)s : {});
+var %(EXPORT_NAME)s%(promise_variable_suffix)s = (%(src)s)(typeof %(EXPORT_NAME)s === 'object' ? %(EXPORT_NAME)s : {});
```

Comment:
Highlighting that this introduces a new variable when using `MODULARIZE_INSTANCE` -- `Module_promise`, or `{EXPORT_NAME}_promise`. This was necessary because we don't want to overwrite the `Module` variable with a Promise, we only want to override it with the actual instance.

Reply:
How about `Module.promise`, a property, instead of a completely new global?

Reply:
This harkens back to the discussion from #5820 where we decided to go with a function that returns a Promise for modularized builds, rather than a making a new property like `ready` or `promise` on `Module`. I think adding a new property just in the `MODULARIZE_INSTANCE` case would be more confusing for users than adding a new global.

If you want to add a property in this case, I think we should go with that design for all cases, even non-modularized ones, removing `onRuntimeInitialized`.

Reply:
Interesting... Well, I admit I don't remember the full discussion there, but maybe let's go back to everyone there and make sure this is what people wanted?

To me it seems kind of awkward to have `Module, Module_promise` globals. We should really have just one global in modularize mode, as one goal is to minimize global stuff. And, `Module_promise` seems like a typo of `Module.promise` to me personally...

Reply:
I agree on the awkwardness. I don't think it's that much more confusing than the existing behavior though -- that `Module` might be a factory function or an instance, depending on the build options. Since this isn't the default behavior, I think it'll be ok with clear docs.

To be clear `Module_promise` variable will only exist for `MODULARIZE_INSTANCE`, when `EXPORT_ES6` isn't used. It will not exist for `MODULARIZE`. Do you have a sense of how much `MODULARIZE_INSTANCE` is used?

Reply:
The terminology has always been awkward. There's always been two "modules" in MODULARIZE mode, which doesn't really modularize, it's a factory!

In regular MODULARIZE mode, you don't get the internal Module until the promise resolves, right? It should be the same for MODULARIZE_INSTANCE.

Reply:
I think I see... are you saying that `Module` doesn't have a value until `Module_promise` resolves? So initially, in `MODULARIZE_INSTANCE`, `Module` is undefined while `Module_promise` is a promise?

How about if it does have an initial value of
```javascript
var Module = {
  promise: ..the value of Module_promise..
};
```
and there is no `Module_promise` global? That seems backwards compatible + avoids a new global, but I may be missing something...

Reply:
Correct. `Module` only has a value if the user has defined it. If it exists it's used as the argument to the factory function, seen [here](https://github.com/emscripten-core/emscripten/blob/master/emcc.py#L3442). If the user never provides it, `Module` is never defined. The only global is `Module_promise`.

I don't think we could provide an initial value for `Module` given that behavior.

We could also change the suffix to `Promise` instead of `_promise` to keep with the camelCasing convention.

Reply:
I'm a little unclear on if that description is about how it's always been, or if it's a new limitation in this PR?

AFAIK, there wasn't such a problem that required a new `Module_promise` before, so I'd guess it's new? If so, why is it needed? Before we also had to be able to access *something*, and it worked somehow?

Reply:
This is a description of the existing behavior. It hasn't been highlighted before because the factory function executed synchronously, so the overwriting of any existing `Module` global variable wasn't as obvious, and accessing properties on it wasn't as difficult.